Sources/MailAgent/MailAgent.php

```diff
@@ -0,0 +1,311 @@
+	 */
+	public function getImplementationClassKeyName(): string
+	{
+		$class_name = get_class($this);
+
+		if ($position = strrpos($class_name, '\\'))
+			return substr($class_name, $position + 1);
+
+		return $class_name;
+	}
+
+	/***********************
+	 * Public static methods
+	 ***********************/
+
+	/**
+	 * Try to load up a supported agent method.
+	 * This is saved in $loaded_api if we are not overriding it.
+	 *
+	 * @todo Add a reference to Utils::$context['instances'] as well?
+	 *
+	 * @param string $overrideCache Allows manually specifying a cache accelerator engine.
+	 * @param bool $fallbackSMF Use the default SMF method if the accelerator fails.
+	 * @return object|false An instance of a child class of this class, or false on failure.
+	 */
+	final public static function load(bool $loadDefault = false)
+	{
+		if (!$loadDefault && !isset(self::$agent))
+		{
+			self::$agent = empty(Config::$modSettings['mail_type']) || Config::$modSettings['smtp_host'] == '' ? 'SendMail' : Config::$modSettings['mail_type'];
+
+			// Handle some other options.
+			if (self::$agent === '1')
+			{
+				self::$agent = 'SMTP';
+			}
+			elseif (self::$agent === '2')
+			{
+				self::$agent = 'SMTPTLS';
+			}
+		}
+
+		if (is_object(self::$loaded_api))
+			return self::$loaded_api;
+		
+		if (is_null(self::$loaded_api))
+			self::$loaded_api = false;
+
+		// What agent we are going to try.
+		$agent_class_name = !empty(self::$agent) ? self::$agent : self::APIS_DEFAULT;
+		$fully_qualified_class_name = self::APIS_NAMESPACE . $agent_class_name;
+
+		// Do some basic tests.
+		$agent_api = false;
+		if (class_exists($fully_qualified_class_name))
+		{
+			$agent_api = new $fully_qualified_class_name();
+
+			// There are rules you know...
+			if (!($agent_api instanceof MailAgentInterface) || !($agent_api instanceof MailAgent))
+				$agent_api = false;
+
+			// No Support?  NEXT!
+			if ($agent_api && !$agent_api->isSupported() && !$agent_api->isConfigured())
+			{
+				// Can we save ourselves?
+				if ($agent_class_name !== self::APIS_DEFAULT)
+					return self::load(true);
+
+				$agent_api = false;
+			}
+
+			// Connect up to the agent.
+			if ($agent_api && $agent_api->connect() === false)
+				$agent_api = false;
+		}
+
+		if (!$agent_api && $agent_class_name !== self::APIS_DEFAULT)
+			$agent_api = self::load(self::APIS_NAMESPACE . self::APIS_DEFAULT, false);
+
+		return $agent_api;
+	}
+
+	/**
+	 * Get the installed Cache API implementations.
+	 */
+	final public static function detect()
+	{
+		$loaded_apis = array();
+
+		$api_classes = new \GlobIterator(self::APIS_FOLDER . '/*.php', \FilesystemIterator::NEW_CURRENT_AND_KEY);
+
+		foreach ($api_classes as $file_path => $file_info)
+		{
+			$class_name = $file_info->getBasename('.php');
+			$fully_qualified_class_name = self::APIS_NAMESPACE . $class_name;
+
+			if (!class_exists($fully_qualified_class_name))
+				continue;
+
+			/* @var CacheApiInterface $cache_api */
```

Comment:
Nit: Here and elsewhere, the documentation needs to have references to CacheApi, etc., changed.

Reply:
Well, guess that shows what I copied from.  Fixed.